pages/blog/typescript-graphql-union-types.mdx

```diff
@@ -0,0 +1,146 @@
+  amount: number
+}
+
+type Food = Sushi | Pizza
+```
+
+[Here's a live example in TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAygrgZwBYEsoF4oG8CwAoKKAaxQDsATALigHJ5kUb9CAzFZahYAJzIHNmUAIYBbAPZxSwaqTgiARhG74AvvnyhIUAAooAXnqEZsgkhWo1dBoUwJRgYsGH6ce-QaIlSZcxcrxqeOp4mtAAYmJi5Mb0qFAAPjr6hsFAA)
+
+And another Union type example, this time with `string` and `string`:
+
+```ts
+type PizzaAddons = 'olives' | 'mushrooms' | 'pepperoni'
+
+const myPreference: PizzaAddons = 'olives' // ok
+const myPreference: PizzaAddons = 'pineapple' // error
+```
+
+[Here's a live example in TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBACglgLwQQwIIBN0HsB2BnKAXigHIsAbOANwjxKgB9SBbAVzwAsAnLLZuxqUhhIPHHBIBYAFAyAxrjzAozEDC4QAZhA045EAFyxEKDNnxFSFarXoB6O1CwBreYuWr1WnRD2HjSGiYipYkYHA4EMgi5BD2jjo8XDJAA)
+
+# GraphQL union and TypeScript union type
+
+GraphQL is a query language that is used to define the API of a server. Like TypeScript or even Go language, GraphQL supports union and interface types. The type system is one of its core principles and we can use it to make type-safe API calls. The goal is to use it to propagate our backend types to the frontend.
+When I started to learn TypeScript, I found myself making lots of type errors and annoying misspellings. These errors became more frequent as the project grew.
+The best solution for that situation was to add type-safety based on a GraphQL schema.
+
+A GraphQL schema can return one of the multiple object types thanks to the Union or Interface keywords. Like TypeScript, we can declare which object types are included in the union:
+
+```graphql
+union Menu = Pizza | Sushi
+```
+
+Let's see a basic example schema that defines a `Menu` union type that can return either a `Pizza` or `Sushi`:
+
+```graphql
+union Menu = Pizza | Sushi
+
+type Pizza {
+  title: String!
+  topping: String
+}
+
+type Sushi {
+  name: String!
+  fish: String
+}
+
+type Query {
+  menu: [Menu!]
+}
+```
+
+After we declare our types in GraphQL and create a union type to return some of the other types, we can generate TypeScript types from the GraphQL schema and use them in the frontend.
+
+# How to use GraphQL Codegen to generate code and deal with TypeScript union types
+
+First of all, you are more than welcome to explore all the plugins of GraphQL Codegen. [GraphQL Code Generator](https://graphql-code-generator.com/)
+Like we did in the previous example, let's talk about getting the data typed safely in the front.
+
+First, let's create our `GetFullMenu` query as follows:
+
+(in this case, I'll use the `react-query` example)
+
+```ts
+const menuQuery = gql`
+  query GetFullMenu {
+    menu {
+      ...Sushi
+      ...Pizza
+    }
+  }
+`
+```
+
+Get the `Menu` type from the schema with the query, so we will be able to get the objects inside the `Menu`.
+
+```ts
+const Menu = () => {
+  const { data } = useQuery<MenuQuery>('menu', async () => {
+    const { menu } = await request(endpoint, menuQuery)
```

Comment:
This example is not 100% clear. Also, I think you should add an example for discriminating the result based on the `__typename` (either using `if` of `switch`)